tensorstore/kvstore/s3/aws_metadata_credential_provider.cc

```diff
@@ -0,0 +1,179 @@
+// Copyright 2023 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <string>
+
+#include "absl/time/time.h"
+#include "absl/strings/str_cat.h"
+#include "tensorstore/kvstore/s3/aws_metadata_credential_provider.h"
+#include "tensorstore/internal/json_binding/absl_time.h"
+#include "tensorstore/internal/json_binding/bindable.h"
+#include "tensorstore/internal/json_binding/json_binding.h"
+
+using ::tensorstore::Result;
+using ::tensorstore::internal::ParseJson;
+using ::tensorstore::internal_http::HttpResponseCodeToStatus;
+using ::tensorstore::internal_kvstore_s3::AwsCredentials;
+
+namespace jb = tensorstore::internal_json_binding;
+
+namespace tensorstore {
+namespace internal_kvstore_s3 {
+
+namespace {
+
+// Token ttl header
+static constexpr char kTokenTtlHeader[] = "x-aws-ec2-metadata-token-ttl-seconds";
+// Token header
+static constexpr char kMetadataTokenHeader[] = "x-aws-ec2-metadata-token";
+// Obtain Metadata server API tokens from this url
+static constexpr char kTokenUrl[] = "http://http://169.254.169.254/latest/api/token";
```

Comment:
This all looks wrong. :)

Reply:
> There looks like some extra http:// there... :)

Ooof yes and in the test cases too. no more late-night coding for me. 
 
> https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
> 
> See also the aws sdk in go:
> 
> * https://github.com/aws/aws-sdk-go/tree/main/aws/ec2metadata

Thanks, I think without the double `http` this lines up with the go implementation. BTW I had a closer look at how the go handles credentials and implemented #119 based on the credential implementation there. Let me know if you want to merge #119 into this branch or if you want to review it separately.

With that, signing off for the night :-)

Reply:
So there are still lots of `http://http://` left in this code.

Reply:
Apologies again. This should now be fixed.